### DISCUSSION TITLE
How access key with space in JsProxy

### DISCUSSION THREAD

**Discussion by janober** [pyodide/pyodide#3537] [General]
Hello,

Really great project! I was surprised by how fast and easy it was to get everything up and running. But sadly can not figure out how to access keys that contain spaces. So if I have a JavaScript object like this:
```
const test = {
  "a": {
    "b b": 1
  }
}
```
Then I can access in Python "a" via `test.a` but how do I access "test['a']['b 'b']"?

Thanks a lot!

**Comment by hoodmane** [comment]
You probably want to use `as_object_map` but you can also use `getattr`.

**Reply by janober** [reply]
Thanks a lot. But do not seem to get it to work. 

The following throws an error:
```
test.as_object_map()["a"]["b b"]
```

The only way I can get it to work is when I convert the direct parent.
```
test.a.as_object_map()["b b"]
```

Is that a bug or intended?

**Reply by hoodmane** [reply]
Currently it's the intended behavior. `as_object_map` is relatively new so we haven't gotten much feedback on it yet. I think it would be reasonable to add a boolean argument `hereditary` which if `True` applies `as_object_map` also to the results of indexing, but probably only if the prototype of the result is `Object` or `null`, but not if it has some more complicated type.  Perhaps we could also allow the user to pass some conditional to determine which objects get wrapped. If we keep the default to `hereditary=False` then this would be a backwards compatible change.

**Reply by janober** [reply]
That sounds great. Also would it be amazing if it would be possible setting it as the default behavior returning everything always like that.

**Reply by hoodmane** [reply]
People have requested a similar thing of a global option to make `dict_converter=Object.fromEntries` the default option so that `to_js` would convert dicts to objects. Currently we don't have any global settings like this. I guess we could add `pyodide.ffi.set_default_dict_converter` for that. I'm not sure what it would look like for `as_object_map`: `object_maps_as_default(True)`?

Of course another option that works already is to use `to_py` to do a deep conversion, but this has a performance cost whereas the wrapper is free:
```py
run_js("""({
	a: 2,
	b: {c : 9}
})""").to_py()["b"]["c"]
```

**Reply by hoodmane** [reply]
We'd have to discuss how we want to style global settings APIs so we can make them reasonably consistent and we have never gotten to it but I am not against it in principle.

**Reply by janober** [reply]
Thanks a lot. Is very helpful to get some insights about how that could work.

To provide you also some insights into why I am interested in it. Which maybe helps you to make a better decision. A manual conversion would def. not be great for my use case. I want to provide a Python Node within [n8n.io](https://n8n.io/) so that people can also write Python code inside of a workflow. That should be as convenient for people as possible and should work as closely to how they expect Python to behave. Currently, we would then either have to:
- tell people to always do the above in such cases (which would cause many frustrations as it is unexpected behavior)
or
- write some kind of additional wrapper Python wrapper for literally every JavaScript Proxy we want to expose (and there are quite a few). So next to the one-time, also long-term support, increased risk of things breaking + additional overhead so reduced speed.

Hope that is helpful!

**Reply by hoodmane** [reply]
Here's a horrible hack that probably does roughly what you want:
```py
from _pyodide_core import jsproxy_typedict, js_flags
from js import Object
Object.new().as_object_map()
jsproxy_typedict[0] = jsproxy_typedict[js_flags['IS_OBJECT_MAP']]
o = run_js("({a : {b: 7}})")
assert o["a"]["b"] == 7
```

**Reply by janober** [reply]
Amazing, thank you very much! Even though it is maybe "horrible" the way it achieves it, but the result is beautiful! 

Thank you very much for your help!  It works great, and will use it temporarily until something "prettier" is available. You def. helped me a lot and also saved me from a lot of work.

**Reply by hoodmane** [reply]
Okay, I think the following is a slightly better version:
```py
from _pyodide_core import jsproxy_typedict
from js import Object
jsproxy_typedict[0] = type(Object.new().as_object_map())
```
I suspect this will keep working for releases in the foreseeable future since this part of the architecture is quite stable. But of course no guarantees.

**Reply by janober** [reply]
That is very appreciated. Did just test. Works great and looks cleaner. Is already changed in the code. 
Thank you very much!
